### PR TITLE
Initial migrating adapter.

### PR DESCRIPTION
**Pull request by jacob-meacham**

### CODE REVIEW

File: .babelrc [hunk 1]
```diff
@@ -0,0 +1,11 @@
+{
+  "presets": ["node6", "stage-1"],
```
Comment (mattsills) [inline .babelrc:2]: so nice not to have to worry about node4.3...

File: package.json [hunk 2]
```diff
@@ -0,0 +1,49 @@
+{
+  "name": "parse-server-migrating-adapter",
+  "version": "0.1.0",
+  "engines": {
+    "node": ">=4.0"
+  },
+  "description": "Parse Server file adapter to migrate between adapters",
+  "author": "jemonjam <jacob.e.meacham@gmail.com>",
+  "license": "MIT",
```
Comment (mattsills) [inline package.json]: This is what we want?

File: package.json [hunk 3]
```diff
@@ -0,0 +1,49 @@
+{
+  "name": "parse-server-migrating-adapter",
+  "version": "0.1.0",
+  "engines": {
+    "node": ">=4.0"
```
Comment (mattsills) [inline package.json]: Do we require node >= 6.0 now?

File: src/index.js [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+  // For a given config object, filename, and data, store a file
+  // Returns a promise
+  createFile(filename, data) {
+    return this.mainAdapter.createFile(filename, data)
+  }
+
+  deleteFile(filename) {
+    // Attempt a delete from all stores:
+    const adapters = [this.mainAdapter].concat(this.oldAdapters)
```
Comment (mattsills) [inline src/index.js:37]: Assert non-empy?

File: src/index.js [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    const promises = adapters.map((adapter) => {
+      return Promise.resolve(adapter.deleteFile(filename))
+    })
+
+    return Promise.settle(promises)
+      .then((results) => {
+        for (const result of results) {
+          if (result.isResolved()) {
+            return Promise.resolve(result.value())
```
Comment (mattsills) [inline src/index.js:46]: Is Promise.any a better fit for this? Or do we care about order?

File: src/index.js [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+      const promises = this.oldAdapters.map((adapter) => {
+        return Promise.resolve(adapter.getFileData(filename))
+      })
+
+      return Promise.settle(promises)
+      .then((results) => {
+        for (const result of results) {
+          if (result.isResolved()) {
+            this.createFile(filename, result.value())
+            return Promise.resolve(result.value())
```
Comment (mattsills) [inline src/index.js]: ditto above

File: package.json (hunk 2 above)
Comment (jacob-meacham) [reply package.json]: good question! Palantir seems to be using Apache. For something as simple as this, I think MIT is fine but it's worth looking into as we start open-sourcing more interesting code

File: src/index.js (hunk 5 above)
Comment (jacob-meacham) [reply src/index.js:46]: I want to enforce that all deletions are done before returning to the caller, which is why I chose settle

File: src/index.js (hunk 6 above)
Comment (jacob-meacham) [reply src/index.js]: This is a better case for Promise.any, made the change

File: src/index.js (hunk 4 above)
Comment (jacob-meacham) [reply src/index.js:37]: done

**Comment by jacob-meacham** [comment]
@mattsills ready for re-review

File: src/index.js [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+        return err
+      })
+    })
+  }
+
+  getFileLocation(config, filename) {
+    return this.mainAdapter.getFileLocation(config, filename)
+  }
+
+  // TODO: Handle file streams
```
Comment (mattsills) [inline src/index.js:75]: What do we prefer, this or `getFileStream(filename) { throw ... }`?

**Review by mattsills — APPROVED** [review]

**Pull request merged**